### ISSUE TITLE
FastlaneCore::DeviceManager.connected_devices is incorrect when using USB hub

### ISSUE THREAD

**Issue by kdubb** [fastlane/fastlane#7675]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

### Issue Description

The `DeviceManager::connected_devices` does not report devices connected to an external USB hub. In [device_manager.rb](https://github.com/fastlane/fastlane/blob/master/fastlane_core/lib/fastlane_core/device_manager.rb#L74l) the comment refers to "looping" in case it has more than 1 controller. 

Unfortunately USB is a tree, for this code to work properly it needs to be a recursive search to find devices connected to a hub, that are connected to a hub, ad infinitum.

##### Complete output when running fastlane, including the stack trace and command used

As this isn't a failure there isn't any stack trace produce.  The `connected_devices` method finishes without exception; it just returns an incorrect list of devices.

### Environment

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                 | Value                                                          |
| ------------------- | -------------------------------------------------------------- |
| OS                  | 10.12.2                                                        |
| Ruby                | 2.0.0                                                          |
| Bundler?            | false                                                          |
| Git                 | git version 2.10.1 (Apple Git-78)                              |
| Installation Source | /usr/local/bin/fastlane                                        |
| Host                | Mac OS X 10.12.2 (16C67)                                       |
| Ruby Lib Dir        | /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib |
| OpenSSL Version     | OpenSSL 0.9.8zc 19 Mar 2015                                    |
| Is contained        | false                                                          |
| Xcode Path          | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version       | 8.2                                                            |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
fastlane_require 'yaml'

workspace = "MessagesKit.xcworkspace"
edge_compositions = {
  'test' => "fastlane/edge-test-compose.yml",
  'dev' => "fastlane/edge-dev-compose.yml"
}


# Build Configuraion

$build = YAML.load(open('../build.yaml'))

before_all do
  # Set environment variables for docker-compose
  container = $build['server']['container']
  ENV['EDGE_CONTAINER_REG'] = container['registry']
  ENV['EDGE_CONTAINER_REPO'] = container['repository']
  ENV['EDGE_CONTAINER_TAG'] = container['tag']
end

# MARK: Test Server Environment

desc "Starts an Edge server environment (Docker composition)"
lane :start_environment do |options|

  type = options[:type] || :dev
  name = options[:name] || type
  pull = options[:pull] || true

  files = [
    edge_compositions[type.to_s]
  ]

  # Pull image (if requested)
  if pull
    docker_compose(
      files: files,
      command: "pull"
    )
  end

  docker_compose(
    files: files,
    command: "up",
    project: name.to_s
  )
  
  port_info = docker_port(
    container: "#{name}_edge_1",
    port: 8080,
  )

  # Ensure we have a publicly usable address
  if port_info[:host] == '0.0.0.0'
    ip = Socket.ip_address_list.detect{|intf| intf.ipv4_private?}
    port_info[:host] = ip.ip_address
  end

  http_service_available(
    target: "http://#{port_info[:host]}:#{port_info[:port]}/admin/healthcheck"
  )

  port_info
end


desc "Stops an Edge server environment (Docker composition)"
lane :stop_environment do |options|

  type = options[:type] || :dev
  name = options[:name] || type

  files = [
    edge_compositions[type.to_s]
  ]

  docker_compose(
    files: files,
    command: "down",
    project: name.to_s
  )
end

# MARK: iOS Tests

platform :ios do

  sim_os_versions = [
    '10.2'
  ]

  sim_device_types = [
    'iPhone 7',
    'iPhone 5'
  ]

  lane :build_and_test do |options|

    # Simulator testing

    # xctest_build(
    #   workspace: workspace,
    #   scheme: "MessagesKit-iOS-Testing",
    #   simulator: true
    # )

    # unit_tests(
    #   sim_os_versions: sim_os_versions,
    #   sim_device_types: sim_device_types
    # )

    # integration_tests(
    #   sim_os_versions: sim_os_versions,
    #   sim_device_types: sim_device_types
    # )


    # Device testing
    devices = FastlaneCore::DeviceManager.connected_devices("iOS")
    if !devices.empty?

      xctest_build(
        workspace: workspace,
        scheme: "MessagesKit-iOS-Testing",
        simulator: false
      )

      devices = devices.map { |d| d.udid }

      unit_tests(
        devices: devices
      )

      integration_tests(
        devices: devices
      )
    
    end

  end

  lane :unit_tests do |options|

    xctest_run(
      **options,
      skip_testing: [
        'MessagesKit-iOSTests/RoundTripTests',
        'MessagesKit-iOSTests/MessagesKitTests'
      ],
      output_type: 'Unit',
      fail_build: false
    )

  end

  lane :integration_tests do |options|
    
    integration_id = (0...8).map { (65 + rand(26)).chr }.join.downcase

    server_info = start_environment(
      type: :test,
      name: integration_id,
      pull: true
    )

    begin

      xctest_run(
        **options,
        launch_args: [
          "-io.retxt.test.Host #{server_info[:host]}",
          "-io.retxt.test.Port #{server_info[:port]}"
        ],
        only_testing: [
          'MessagesKit-iOSTests/RoundTripTests',
          'MessagesKit-iOSTests/MessagesKitTests'
        ],
        output_type: 'Integration',
        fail_build: false
      )

    ensure

      stop_environment(
        type: :test,
        name: integration_id
      )

    end

  end

end

# MARK: macOS Tests

platform :mac do

  lane :build_and_test do |options|

    xctest_build(
      workspace: workspace,
      scheme: "MessagesKit-macOS-Testing",
      code_coverage: true
    )
    
    # unit_tests

    integration_tests
    
  end

  lane :unit_tests do |options|
    
    xctest_run(
      archs: [
        'x86_64'
      ],
      skip_testing: [
        'MessagesKit-macOSTests/RoundTripTests',
        'MessagesKit-macOSTests/MessagesKitTests'
      ],
      output_type: 'Unit',
      fail_build: false
    )
        
  end

  lane :integration_tests do |options|
    
    integration_id = (0...8).map { (65 + rand(26)).chr }.join.downcase

    server_info = start_environment(
      type: :test,
      name: integration_id,
      pull: true
    )

    begin

      xctest_run(
        archs: [
          'x86_64'
        ],
        launch_args: [
          "-io.retxt.test.Host #{server_info[:host]}",
          "-io.retxt.test.Port #{server_info[:port]}"
        ],
        only_testing: [
          'MessagesKit-macOSTests/RoundTripTests',
          'MessagesKit-macOSTests/MessagesKitTests'
        ],
        output_type: 'Integration',
        fail_build: false
      )

    ensure
      
      stop_environment(
        type: :test,
        name: integration_id
      )

    end

  end

end
```
</details>

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.3.0   | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| slack-notifier            | 1.5.1        |
| rouge                     | 1.11.1       |
| xcpretty                  | 0.2.4        |
| terminal-notifier         | 1.7.1        |
| unicode-display_width     | 1.1.2        |
| terminal-table            | 1.7.3        |
| plist                     | 3.2.0        |
| public_suffix             | 2.0.4        |
| addressable               | 2.5.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| babosa                    | 1.0.2        |
| colored                   | 1.2          |
| highline                  | 1.7.8        |
| commander                 | 4.4.2        |
| excon                     | 0.54.0       |
| faraday                   | 0.10.0       |
| unf_ext                   | 0.0.7.2      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20161129 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| fastimage                 | 2.0.1        |
| gh_inspector              | 1.0.2        |
| uber                      | 0.0.15       |
| representable             | 2.3.0        |
| retriable                 | 2.1.0        |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| hurley                    | 0.2          |
| little-plugger            | 1.1.4        |
| multi_json                | 1.12.1       |
| logging                   | 2.1.0        |
| jwt                       | 1.5.6        |
| memoist                   | 0.15.0       |
| os                        | 0.9.6        |
| signet                    | 0.7.3        |
| googleauth                | 0.5.1        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.9.20       |
| mini_magick               | 4.5.1        |
| multi_xml                 | 0.6.0        |
| rubyzip                   | 1.2.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 0.0.4        |
| dotenv                    | 2.1.1        |
| bundler                   | 1.13.6       |
| faraday_middleware        | 0.10.1       |
| i18n                      | 0.7.0        |
| thread_safe               | 0.3.5        |
| tzinfo                    | 1.2.2        |
| minitest                  | 5.10.1       |
| activesupport             | 4.2.7.1      |
| json                      | 1.8.3        |
| CFPropertyList            | 2.3.4        |
| claide                    | 1.0.1        |
| nanaimo                   | 0.2.3        |
| xcodeproj                 | 1.4.2        |
</details>

*generated on:* **2016-12-26**
</details>

**Comment by mfurtak** [comment]
Thanks for letting us know! If you have example output from the `system_profiler` for this case as well, that would be awesome!

**Comment by mfurtak** [comment]
Never mind on the above - I captured my own output. 👍

**Comment by mfurtak** [comment]
This has been addressed with #7727 and will go out with the next release - thanks! 🚀

**Comment by mfurtak** [comment]
This was released as part of [_fastlane_ 2.5.0](https://github.com/fastlane/fastlane/releases/tag/2.5.0) 🎉
